<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Проект-«Предобработка-данных»" data-toc-modified-id="Проект-«Предобработка-данных»-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Проект «Предобработка данных»</a></span></li><li><span><a href="#Исследование-надёжности-заёмщиков" data-toc-modified-id="Исследование-надёжности-заёмщиков-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Исследование надёжности заёмщиков</a></span><ul class="toc-item"><li><span><a href="#Открытие-данных.--" data-toc-modified-id="Открытие-данных.---2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Открытие данных.  <a id="id_1"></a></a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Предобработка-данных--" data-toc-modified-id="Предобработка-данных---3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Предобработка данных  <a id="2"></a></a></span><ul class="toc-item"><li><span><a href="#Обработка-пропусков" data-toc-modified-id="Обработка-пропусков-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Обработка пропусков</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-3.1.1"><span class="toc-item-num">3.1.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Замена-типа-данных" data-toc-modified-id="Замена-типа-данных-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Замена типа данных</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-3.2.1"><span class="toc-item-num">3.2.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Обработка-дубликатов" data-toc-modified-id="Обработка-дубликатов-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Обработка дубликатов</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-3.3.1"><span class="toc-item-num">3.3.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li></ul></li><li><span><a href="#Лемматизация" data-toc-modified-id="Лемматизация-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Лемматизация</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Категоризация-данных" data-toc-modified-id="Категоризация-данных-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Категоризация данных</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Ответы-на-вопросы" data-toc-modified-id="Ответы-на-вопросы-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Ответы на вопросы</a></span><ul class="toc-item"><li><span><a href="#Есть-ли-зависимость-между-наличием-детей-и-возвратом-кредита-в-срок?" data-toc-modified-id="Есть-ли-зависимость-между-наличием-детей-и-возвратом-кредита-в-срок?-6.1"><span class="toc-item-num">6.1&nbsp;&nbsp;</span>Есть ли зависимость между наличием детей и возвратом кредита в срок?</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-6.1.1"><span class="toc-item-num">6.1.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Есть-ли-зависимость-между-семейным-положением-и-возвратом-кредита-в-срок?" data-toc-modified-id="Есть-ли-зависимость-между-семейным-положением-и-возвратом-кредита-в-срок?-6.2"><span class="toc-item-num">6.2&nbsp;&nbsp;</span>Есть ли зависимость между семейным положением и возвратом кредита в срок?</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-6.2.1"><span class="toc-item-num">6.2.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Есть-ли-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок?" data-toc-modified-id="Есть-ли-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок?-6.3"><span class="toc-item-num">6.3&nbsp;&nbsp;</span>Есть ли зависимость между уровнем дохода и возвратом кредита в срок?</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-6.3.1"><span class="toc-item-num">6.3.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Как-разные-цели-кредита-влияют-на-его-возврат-в-срок?" data-toc-modified-id="Как-разные-цели-кредита-влияют-на-его-возврат-в-срок?-6.4"><span class="toc-item-num">6.4&nbsp;&nbsp;</span>Как разные цели кредита влияют на его возврат в срок?</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-6.4.1"><span class="toc-item-num">6.4.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li></ul></li><li><span><a href="#Общий-вывод" data-toc-modified-id="Общий-вывод-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Общий вывод</a></span></li><li><span><a href="#Чек-лист-готовности-проекта" data-toc-modified-id="Чек-лист-готовности-проекта-8"><span class="toc-item-num">8&nbsp;&nbsp;</span>Чек-лист готовности проекта</a></span></li></ul></div>

# Проект «Предобработка данных»

# Исследование надёжности заёмщиков

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.

Результаты исследования будут учтены при построении модели **кредитного скоринга** — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

## Открытие данных.  <a id="id_1"></a>

In [1]:
import pandas as pd
from pymystem3 import Mystem
from collections import Counter

df = pd.read_csv('/datasets/data.csv')
df.head(5)

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

In [2]:
df.tail(5)

children  days_employed  dob_years education  education_id  \
21520         1   -4529.316663         43   среднее             1   
21521         0  343937.404131         67   среднее             1   
21522         1   -2113.346888         38   среднее             1   
21523         3   -3112.481705         38   среднее             1   
21524         2   -1984.507589         40   среднее             1   

          family_status  family_status_id gender income_type  debt  \
21520  гражданский брак                 1      F   компаньон     0   
21521   женат / замужем                 0      F   пенсионер     0   
21522  гражданский брак                 1      M   сотрудник     1   
21523   женат / замужем                 0      M   сотрудник     1   
21524   женат / замужем                 0      F   сотрудник     0   

        total_income                       purpose  
21520  224791.862382             операции с жильем  
21521  155999.806512          сделка с автомобилем  
21522   89672.561153                  недвижимость  
21523  244093.050500  на покупку своего автомобиля  
21524   82047.418899         на покупку автомобиля

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


In [4]:
df.describe()

children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  140827.311974     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt  total_income  
count      21525.000000  21525.000000  1.935100e+04  
mean           0.972544      0.080883  1.674223e+05  
std            1.420324      0.272661  1.029716e+05  
min            0.000000      0.000000  2.066726e+04  
25%            0.000000      0.000000  1.030532e+05  
50%            0.000000      0.000000  1.450179e+05  
75%            1.000000      0.000000  2.034351e+05  
max            4.000000      1.000000  2.265604e+06

## Вывод

После первоначального ознакомления с данными можно уже увидеть что нужно будет сделать на этапе предобработки данных, есть пропуски, проблемы с регистром, отрицательные значения, где их быть не должно, наименования столбцов тоже не мешало бы причесать для удобства. Типы данных тоже следует привести в порядок. 

#  Предобработка данных  <a id="2"></a>
-Обработка пропусков  
-Дубликатов  
-Замена типов данных  

## Обработка пропусков

Получим список столбцов и переименуем их

In [5]:
df.columns.values.tolist() 



['children',
 'days_employed',
 'dob_years',
 'education',
 'education_id',
 'family_status',
 'family_status_id',
 'gender',
 'income_type',
 'debt',
 'total_income',
 'purpose']

Переименуем столбцы

In [6]:
df.set_axis(['children','days_employed', 'client_years', 
             'education', 'education_id', 'family_status', 
             'family_status_id', 'gender', 'client_category',
             'loan_return_status', 'income', 'purpose'], axis = 'columns', inplace = True)


In [7]:
df.columns.values.tolist()

['children',
 'days_employed',
 'client_years',
 'education',
 'education_id',
 'family_status',
 'family_status_id',
 'gender',
 'client_category',
 'loan_return_status',
 'income',
 'purpose']

Определим и заполним пропущенные значения.

In [8]:
df.isnull().sum()

children                 0
days_employed         2174
client_years             0
education                0
education_id             0
family_status            0
family_status_id         0
gender                   0
client_category          0
loan_return_status       0
income                2174
purpose                  0
dtype: int64

Количество пропусков в двух столбцах идентичны. Надо понять почему. У нас пропуски в стаже и в уровне дохода. Пустые значения Nan в доходе и в стаже работы - значит скорее всего эти люди не работали и не имеют дохода, т.к. количество точно совпадает в обеих категориях. 





In [9]:
df['days_employed'] = df['days_employed'].fillna(0) #меняем пропуски на нули в 'days_employed'

In [10]:
df['income'] = df['income'].fillna(0) #меняем пропуски на нули в 'income'

In [11]:
df.isnull().sum() #чек

children              0
days_employed         0
client_years          0
education             0
education_id          0
family_status         0
family_status_id      0
gender                0
client_category       0
loan_return_status    0
income                0
purpose               0
dtype: int64

### Вывод

Пропуски заполнили нулями. 

Следует изменить форму заполнения (получения) данных клиентов, чтобы в таких случаях как у нас (люди без стажа и дохода) в них выставлялись значения 0 - для будущих исследований и анализа. Эту инфу следует донести до тех, кто работает над этой формой в банке (через тим лида/руководство/самому (в зависимости от ситуации)), чтобы улучшить процессы внутри компании.

Почему заполняем нулями? Потому-что это наиболее логичный метод в данной ситуации, который не будет искажать результаты исследований.

## Замена типа данных


Приведём в порядок типы данных:

In [12]:
df.dtypes #посмотрим что у нас есть

children                int64
days_employed         float64
client_years            int64
education              object
education_id            int64
family_status          object
family_status_id        int64
gender                 object
client_category        object
loan_return_status      int64
income                float64
purpose                object
dtype: object

2 столбца имеют формат float64 , превратим их целочисленные значения. 

In [13]:
df['days_employed'] = df['days_employed'].astype('int') 
df['income'] = df['income'].astype('int') 
df.dtypes #проверка

children               int64
days_employed          int64
client_years           int64
education             object
education_id           int64
family_status         object
family_status_id       int64
gender                object
client_category       object
loan_return_status     int64
income                 int64
purpose               object
dtype: object

### Вывод

Для изменения типов данных использовался метод .astype для приведения данных в целочисленные значения, для удобства в дальнейшем исследовании. Т.к. данные после запятой очень незначительны, влиянием их на исследование можно пренебречь. 

## Обработка дубликатов

In [14]:
df.duplicated().sum()

54

На данном этапе обработки данных 54 дубля. Хорошо бы понять, почему они появились. 

У нас нет конкретных данных по ФИО клиентов, поэтому непонятно, дубли являются тем же самым лицом или это разные люди. 
Вопросы к тому как собираются данные, и как они обрабатываются. Будем считать это ошибкой выгрузки данных и удалим дубли. 

In [15]:
df = df.drop_duplicates().reset_index(drop=True) #дропаем

In [16]:
df.duplicated().sum() #чек

0

Дубли удалены. 

Теперь перейдём к столбцам с регистром. 
На этапе ознакомления с данными, сразу бросилось в глаза, что некоторые значения в 'education' указаны капсом, хорошо будет преобразовать их в нижний регистр. И для удобства в дальнейшей работе и нашего спокойствия, можно преобразовать вообще все столбцы с текстовыми обозначениями/категориями к единому регистру.

In [17]:
df['education'] = df['education'].str.lower() 
df['family_status'] = df['family_status'].str.lower()
df['client_category'] = df['client_category'].str.lower()
df['purpose'] = df['purpose'].str.lower()

In [18]:
df.head(5) #посмотрим что получилось

children  days_employed  client_years education  education_id  \
0         1          -8437            42    высшее             0   
1         1          -4024            36   среднее             1   
2         0          -5623            33   среднее             1   
3         3          -4124            32   среднее             1   
4         0         340266            53   среднее             1   

      family_status  family_status_id gender client_category  \
0   женат / замужем                 0      F       сотрудник   
1   женат / замужем                 0      F       сотрудник   
2   женат / замужем                 0      M       сотрудник   
3   женат / замужем                 0      M       сотрудник   
4  гражданский брак                 1      F       пенсионер   

   loan_return_status  income                     purpose  
0                   0  253875               покупка жилья  
1                   0  112080     приобретение автомобиля  
2                   0  145885               покупка жилья  
3                   0  267628  дополнительное образование  
4                   0  158616             сыграть свадьбу

Всё ок, идём дальше. 

С последнего взгляда в таблицу, глаз режут эти минусы в графе 'days_employed', самое время разобраться с ними. Понять, почему это произошло, как это решить и как привести данные в порядок для исследования.


In [19]:
#для начала посмотрим диапазон значений и попробуем разобраться
df['days_employed'].max()

401755

In [20]:
df['days_employed'].min()

-18388

In [21]:
df['days_employed'].median()

-989.0

In [22]:
df['days_employed'].mean()

56821.71249592474

Данные очень странные, читать их неудобно, нет наглядности, например непонятно 401755 дней (максимальное значение) - сколько стажа это хотя бы в месяцах, или годах. 

Также второй вопрос возникает: - каким образом могли появиться отрицательные значения в такой категории данных как стаж? 

Возможно дело в заполнении датафрейма или в выгрузке, у меня нет данных о том каким способом собирались данные, может при внесении данных электронным способом, некоторые из менеджеров банка ставили символ "-" перед цифрой, при вводе данных клиента. Но какой клиент сообщает о своём стаже в днях? Совершеннейшая путанница. Возможно клиенты давали свои сведения стандартно, в годах, а в датафрейм после сбора данных производилась конвертация. 

Сейчас я вижу следующий вариант использования имеющихся данных - конвертация значений из дней в годы, для удобства и наглядности, а с минусами: - примем отрицательные значения за таковые без знака минус, т.е. преобразуем их в плюс. 

In [23]:
#убираем минусы
df['days_employed'] = df['days_employed'].abs()

In [24]:
#конвертируем дни в годы
df['years_employed'] = df['days_employed'] / 365
df['years_employed'] = df['years_employed'].round(1)

In [25]:
#удалим столбец с днями
del df['days_employed']

In [26]:
#смотрим что получилось
df.head()

children  client_years education  education_id     family_status  \
0         1            42    высшее             0   женат / замужем   
1         1            36   среднее             1   женат / замужем   
2         0            33   среднее             1   женат / замужем   
3         3            32   среднее             1   женат / замужем   
4         0            53   среднее             1  гражданский брак   

   family_status_id gender client_category  loan_return_status  income  \
0                 0      F       сотрудник                   0  253875   
1                 0      F       сотрудник                   0  112080   
2                 0      M       сотрудник                   0  145885   
3                 0      M       сотрудник                   0  267628   
4                 1      F       пенсионер                   0  158616   

                      purpose  years_employed  
0               покупка жилья            23.1  
1     приобретение автомобиля            11.0  
2               покупка жилья            15.4  
3  дополнительное образование            11.3  
4             сыграть свадьбу           932.2

Взглянем на получившиеся данные в 'years_employed'

In [27]:
df['years_employed'].max()

1100.7

В максимальных у нас как выяснилось неадекватное число.

Полезно понять сколько у нас значений, существенно выходящих за рамки приемлимых и нормальных, т.е. заведомо нереалистичных данных, и понять что у них общего, возможно в этом есть какая-то закономерность, позволяющая разобраться с этими данными. 



In [28]:
df['years_employed'].mean()

165.22554142797262

Из-за максимального значения у нас сбито и среднее значение, тут может помочь медиана.

In [29]:
df['years_employed'].median()

5.0

Медианное значение выглядит неплохо, около пяти лет, вполне себе нормальное значение для трудового стажа в данной выборке. Т.к. в вопросах поставленных перед исследованием, нет вопросов касающихся длительности стажа клиентов - значения столбца можно оставить как есть.

Минимумы смотреть не стал, потому-что использована функция .abs и изменены все значения в фактическое значение ввода без учета знака. С учётом того, что ранее мною были заполнены пропуски на нули в данном столбце - можно сразу сделать вывод, что минимальное значение будет равно нулю.



In [30]:
df.duplicated().sum() 

17

In [31]:
df = df.drop_duplicates().reset_index(drop=True) 

Мы же сделали выравнивание по нижнему регистру, и скорее всего некоторые из выравненных строк тоже были дублями. Удаляем.

In [32]:
df.duplicated().sum() #проверяем

0

### Вывод

Дубли удалены плюс проведена работа с приведением текстовых значений к нижнему регистру и произведена конвертация стажа работы клиентов из дней в годы, а также избавились от отрицательных значений.

# Лемматизация

Взглянем на каждое значение в целях кредита:

In [33]:
df['purpose'].unique() 

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

In [34]:
m = Mystem() #здесь проведём лемматизацию и добавим новый столбец, он нам пригодится в будущем.
df['main_purpose'] = df['purpose'].apply(m.lemmatize) 
df.sample(3)


children  client_years education  education_id     family_status  \
14220         0            52   среднее             1   женат / замужем   
20495         0            33   среднее             1   женат / замужем   
7255          4            35   среднее             1  гражданский брак   

       family_status_id gender client_category  loan_return_status  income  \
14220                 0      F       сотрудник                   0  206561   
20495                 0      M       пенсионер                   0  163411   
7255                  1      M       сотрудник                   0  157568   

                    purpose  years_employed                      main_purpose  
14220          недвижимость            29.4                [недвижимость, \n]  
20495           образование           913.4                 [образование, \n]  
7255   покупка своего жилья             2.7  [покупка,  , свой,  , жилье, \n]

## Вывод

Первое что бросается в глаза, среди целей часто встречаются связанные с недвижимостью. 
Их можно выделить в отдельную категорию: - "недвижимость", далее: "авто", "образование", "свадьба".  




# Категоризация данных

In [35]:
df['purpose'].unique() #взглянем на цели кредита еще раз

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

Есть одно значение не связанное с приобретением недвижимости: - "ремонт жилью" - т.е. цель отличается от преобладающей в массе "покупка+недвижимость", но т.к. такое значение лишь одно, стоит проверить, как часто оно встречается среди всех строк и стоит ли его выделять в отдельную категорию, как "ремонт"

In [36]:
df[df['purpose'] == 'ремонт жилью']['purpose'].count() #посчитаем количество значений категории


607

Всего 607, значений совсем немного от общего количества, явно не требуется выделять в отдельную категорию. Таким образом мы узнали, что это не был отдельный банковский продукт, в духе "кредит для ремонта жилья" который мог быть крупной отдельной категорией. 
Можем спокойно переходить к следующим столбцам и дообработать оставшиеся данные перед исследованием.

In [37]:
df['children'].value_counts() #взглянем на показатели в графе дети

 0     14091
 1      4808
 2      2052
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

47 клиентов с минус одним ребёнком. Тут скорее всего ошибка со знаком.
И у 76 клиентов - по 20 детей. Это точно из областей фантастики, если бы их было 1-2 человека, можно было бы поинтересоваться возрастом, но здесь их слишком много, в современном мире это не распространено совершенно, да и отсутствие показателей свыше 5 детей говорит о многом. 
Будем считать, что это неведомо каким образом появившийся ноль к значению "2".  

In [38]:
df['children'] = df['children'].abs()

In [39]:
df['children'] = df['children'].replace(20, 2) #20 на 2

In [40]:
df['children'].value_counts() #чекаем и идём дальше.

0    14091
1     4855
2     2128
3      330
4       41
5        9
Name: children, dtype: int64

In [41]:
df['client_years'].value_counts() #переходим к следующему столбцу

35    616
40    607
41    605
34    601
38    597
42    596
33    581
39    572
31    559
36    554
44    545
29    544
30    537
48    536
37    536
50    513
43    512
32    509
49    508
28    503
45    496
27    493
52    484
56    483
47    477
54    476
46    472
53    459
57    456
58    454
51    446
55    443
59    443
26    408
60    374
25    357
61    354
62    348
63    269
24    264
64    260
23    252
65    193
22    183
66    182
67    167
21    111
0     101
68     99
69     85
70     65
71     56
20     51
72     33
19     14
73      8
74      6
75      1
Name: client_years, dtype: int64

В целом всё нормально, только смущает 101 значение с возрастом "0" (заменим на медианное). Почему такой вывод? Потому что все остальные значения выглядят нормально и скорее всего, здесь просто отсутствуют данные о возрасте. Замена значения "0" на медианное, вместо удаления данных строк, позволит включить данные строки в анализ, при этом т.к. вопросов напрямую связанных с возрастом в задаче не задано, то такие данные подойдут для ответа на поставленные вопросы. 

In [42]:
df['client_years'].median() #узнаем медианное значение

42.0

In [43]:
df['client_years'] = df['client_years'].replace(0, 43) #заменим 0 на медианное значение


In [44]:
df['client_years'].value_counts() #смотрим что получилось, 43ёхлетних стало на сотку больше), медиана не поменялась, ок идём дальше

35    616
43    613
40    607
41    605
34    601
38    597
42    596
33    581
39    572
31    559
36    554
44    545
29    544
30    537
37    536
48    536
50    513
32    509
49    508
28    503
45    496
27    493
52    484
56    483
47    477
54    476
46    472
53    459
57    456
58    454
51    446
55    443
59    443
26    408
60    374
25    357
61    354
62    348
63    269
24    264
64    260
23    252
65    193
22    183
66    182
67    167
21    111
68     99
69     85
70     65
71     56
20     51
72     33
19     14
73      8
74      6
75      1
Name: client_years, dtype: int64

Взглянем на столбец с образованием:

In [45]:
df['education'].value_counts() 

среднее                15172
высшее                  5250
неоконченное высшее      744
начальное                282
ученая степень             6
Name: education, dtype: int64

Всё как в жизни. Людей с учёной степенью крайне мало от всего числа людей.

In [46]:
df[df['education_id'] == 1].groupby('education')['education'].count() 
#посмотрим чему соответствуют значения 1 и 0 в графе education_id


education
среднее    15172
Name: education, dtype: int64

In [47]:
df[df['education_id'] == 0].groupby('education')['education'].count()


education
высшее    5250
Name: education, dtype: int64

Значения в единицу - среднее образование, а нули - высшее. 
Остальные ребята не участвуют в данной категоризации. Для исследования это не важно, вопросов возвратности в зависимости от наличия высшего образования у клиентов не поставлено, идём дальше.

In [48]:
df['family_status'].value_counts() #здесь всё хорошо

женат / замужем          12339
гражданский брак          4151
не женат / не замужем     2810
в разводе                 1195
вдовец / вдова             959
Name: family_status, dtype: int64

In [49]:
df['family_status_id'].value_counts() 

0    12339
1     4151
4     2810
3     1195
2      959
Name: family_status_id, dtype: int64

Здесь тоже полный порядок и по идентичным цифрам как и в столбце 'family_status' видно, что чему соответствует:  
0 - в официальном браке,  
1 - гражданский брак,  
2 - не в браке,  
3 - в разводе,  
4 - овдовевшие.

In [50]:
df['gender'].value_counts() 

F      14174
M       7279
XNA        1
Name: gender, dtype: int64

Первое что бросается в глаза - женщин в 2 раза больше чем мужчин, и некий XNA - всего 1 значение.

In [51]:
df[df['gender'] == 'XNA']['gender'] #номер строки 10684 - почти идеально в середине, возможно это что-то техническое.

10684    XNA
Name: gender, dtype: object

In [52]:
df.drop([10684], inplace=True) #дропаем

In [53]:
df['gender'].value_counts() #смотрим, ок. 

F    14174
M     7279
Name: gender, dtype: int64

In [54]:
df['client_category'].value_counts() #всё ок

сотрудник          11084
компаньон           5077
пенсионер           3829
госслужащий         1457
предприниматель        2
безработный            2
в декрете              1
студент                1
Name: client_category, dtype: int64

In [55]:
df['loan_return_status'].value_counts() #здесь всё нормально, каких то неожиданных значений нет

0    19712
1     1741
Name: loan_return_status, dtype: int64

In [56]:
df['income'].max() #посмотрим на разные значения доходов

2265604

In [57]:
df['income'].min() #нули мы ранее проставили людям без зарплаты / трудового стажа

0

In [58]:
df['income'].mean() #cреднее значение

151008.0352398266

In [59]:
df['income'].median() #медиана

135773.0

Вроде всё нормально с цифрами, непонятна валюта, т.к. цифры значительно превосходят средние значения по РФ. Будет считать что это Казахстан :) (потому-что свадеб много) там любят празднества ("тои" - как они говорят). 

Выполним категоризацию клиентов по наличию детей.

In [60]:
def clients_group(children):
        """
        Возвращает группу по наличию детей children, используя правила:
        - 'бездетные' при значении children == 0 
        - 'есть дети' при значениии children больше 0
        """

        if children == 0:
                return 'бездетные'
        if children > 0:
                return 'есть дети'
        
df['clients_group'] = df['children'].apply(clients_group)
df['clients_group'].head(10) #чекнем (работает)


0    есть дети
1    есть дети
2    бездетные
3    есть дети
4    бездетные
5    бездетные
6    бездетные
7    бездетные
8    есть дети
9    бездетные
Name: clients_group, dtype: object

Категоризируем клиентов по уровню доходов.

In [61]:
df['income'].quantile([0.25,0.5,0.75]) #для категоризации цифровых значений, таких как зарплаты, удобно будет использовать квантили 

0.25     89088.0
0.50    135773.0
0.75    195799.0
Name: income, dtype: float64

In [62]:
def income_quartiles(income): #придумаем имена квартилям
        """
        Возвращает группу по значению дохода income, используя правила:
        - 'Низкий доход' при значении income <= 89090.0 
        - 'Средний доход при значении income <= 135789.0
        - 'Высокий доход' при значении income < 195818.0
        - 'Сверхвысокий доход' в остальных случаях
        """

        if income <= 88090.0:
                return 'Низкий доход'
        if income <= 135789.00:
                return 'Средний доход'
        if income < 195818.0:
                return 'Высокий доход'
        return 'Сверхвысокий доход'
        
df['income_quartiles'] = df['income'].apply(income_quartiles)
df['income_quartiles'].head(10) #чекнем (работает)


0    Сверхвысокий доход
1         Средний доход
2         Высокий доход
3    Сверхвысокий доход
4         Высокий доход
5    Сверхвысокий доход
6    Сверхвысокий доход
7         Высокий доход
8         Средний доход
9         Высокий доход
Name: income_quartiles, dtype: object

## Вывод

В некоторых столбцах требовались небольшие корректировки, каждое действие подробно описывал, как и причину произведённых изменений. 
Также была выполнена категоризация данных в столбцах имеющих значения для ответов на поставленные вопросы.
Наши данные готовы к выполнению задач и с помощью них мы можем получить ответы на вопросы. 

# Ответы на вопросы

-Есть ли зависимость между наличием детей и возвратом кредита в срок?  
-Есть ли зависимость между семейным положением и возвратом кредита в срок?  
-Есть ли зависимость между уровнем дохода и возвратом кредита в срок?  
-Как разные цели кредита влияют на его возврат в срок?  
<a id="id_3"></a>

## Есть ли зависимость между наличием детей и возвратом кредита в срок?


Для ответа на вопрос посмотрим процент вернувших кредиты в каждом из столбцов, тогда можно будет сделать вывод, основываясь на полученных данных.

In [63]:
children_pivot = df.pivot_table(index=['clients_group'], columns=['loan_return_status'], values='client_years', aggfunc='count')
children_pivot['ratio'] = children_pivot[1] / children_pivot[0] * 100
children_pivot.sort_values(by = 'ratio', ascending = False)
#сделаем сводную таблицу и посмотрим соотношение


loan_return_status      0     1      ratio
clients_group                             
есть дети            6685   678  10.142109
бездетные           13027  1063   8.159975

### Вывод

Четко видно что заёмщики имеющие детей более платёжеспособные. 

## Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [64]:
family_returns = df.pivot_table(index=['family_status'], columns = ['loan_return_status'], values='client_years', aggfunc='count')
family_returns['ratio'] = family_returns[1] / family_returns [0] * 100
family_returns.sort_values(by = 'ratio', ascending = False) #используем аналогичную сводную таблицу как и в предыдущем решении 



loan_return_status         0    1      ratio
family_status                               
не женат / не замужем   2536  274  10.804416
гражданский брак        3762  388  10.313663
женат / замужем        11408  931   8.160940
в разводе               1110   85   7.657658
вдовец / вдова           896   63   7.031250

### Вывод

Есть зависимость, хоть и не прямая. 

Например необязательно состоящие в браке люди - платежеспособнее, чем не состоящие в браке, согласно полученным данным мы видим, что соотношение возвратов имеет близкие значения среди не женатых и среди тех, кто находится в гражданском браке. 

В тоже время женатые люди (в официальном браке) - имеют ниже соотношение, как и те, кто находятся в разводе или овдовели. У них еще более низкое соотношение.

В итоге топ по платежеспособности выглядит так:
1. не женат / не замужем - 10.80
2. гражданский брак - 10.28
3. женат / замужем - 8.15
4. в разводе - 7.65
5. вдовец / вдова - 7.03

Эммм, получается стоит быть очень внимательными при выдаче займов овдовевшим клиентам и тщательно взвешивать все риски.

## Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [65]:
income_returns = df.pivot_table(index=['income_quartiles'], columns = ['loan_return_status'], values='client_years', aggfunc='count')
income_returns['ratio'] = income_returns[1] / income_returns [0] * 100
income_returns.sort_values(by = 'ratio', ascending = False) #используем аналогичную сводную таблицу как и в предыдущем решении 

loan_return_status     0    1     ratio
income_quartiles                       
Высокий доход       4881  481  9.854538
Средний доход       5006  465  9.288853
Низкий доход        4845  412  8.503612
Сверхвысокий доход  4980  383  7.690763

### Вывод

Со всеми значениями закономерная ситуация, чем выше доход - тем выше соотношение платежеспособности и возвратности.
Однако не всё так однозначно со сверхвысокими значениями - возможно, это связано с тем, что неплатежеспособные люди, остро желающие заполучить кредит, будучи не очень платежеспособными - каким-то образом завышают свои доходы, чтобы получить кредит, который в дальнейшем им сложно погасить и мы видим такую печальную статистику возвратов в категории людей с очень высокими доходами.

Ответ на вопрос: Есть ли зависимость между уровнем дохода и возвратом кредита в срок? - Однозначно - да, есть. 

## Как разные цели кредита влияют на его возврат в срок?

In [66]:
df['main_purpose'].value_counts() #взглянем на наши категории после лемматизации 

[автомобиль, \n]                                          972
[свадьба, \n]                                             791
[на,  , проведение,  , свадьба, \n]                       768
[сыграть,  , свадьба, \n]                                 765
[операция,  , с,  , недвижимость, \n]                     675
[покупка,  , коммерческий,  , недвижимость, \n]           661
[операция,  , с,  , жилье, \n]                            652
[покупка,  , жилье,  , для,  , сдача, \n]                 651
[операция,  , с,  , коммерческий,  , недвижимость, \n]    650
[покупка,  , жилье, \n]                                   646
[жилье, \n]                                               646
[покупка,  , жилье,  , для,  , семья, \n]                 638
[строительство,  , собственный,  , недвижимость, \n]      635
[недвижимость, \n]                                        633
[операция,  , со,  , свой,  , недвижимость, \n]           627
[строительство,  , жилой,  , недвижимость, \n]            624
[покупка

In [67]:
#сгруппируем данные
def loan_type(main_purpose): #
        """
        Возвращает группу по типу назначения кредита main_purpose, используя правила:
        - 'Автомобиль' при наличии в строке сочетания букв 'Авто' 
        - 'Образование' при наличии в строке сочетания букв 'Образование'
        - 'Свадьба' при наличии в строке сочетания букв 'Свадьба'
        - 'Недвижимость' при наличии в строке сочетания букв 'недвижимость' или 'жил'
        """

        if 'автомобиль' in main_purpose:
                return 'Автомобиль'
        if 'образование' in main_purpose:
                return 'Образование'
        if 'свадьба' in main_purpose:
                return 'Свадьба'
        if 'недвижимость' or 'жил' in main_purpose:
                return 'Недвижимость'
        
            
df['loan_type'] = df['main_purpose'].apply(loan_type)
df['loan_type'].value_counts()



Недвижимость    10810
Автомобиль       4306
Образование      4013
Свадьба          2324
Name: loan_type, dtype: int64

In [68]:
loan_type_returns = df.pivot_table(index=['loan_type'], columns = ['loan_return_status'], values='client_years', aggfunc='count')
loan_type_returns['ratio'] = loan_type_returns[1] / loan_type_returns [0] * 100
loan_type_returns.sort_values(by = 'ratio', ascending = False) #используем аналогичную сводную таблицу как и в предыдущем решении 

loan_return_status      0    1      ratio
loan_type                                
Автомобиль           3903  403  10.325391
Образование          3643  370  10.156464
Свадьба              2138  186   8.699719
Недвижимость        10028  782   7.798165

### Вывод

Как разные цели кредита влияют на его возврат в срок?

Итак, что мы видим. Цели кредита на его возвратность влияют следующим образом: - займы взятые на обучение и покупку авто имеют более высокое соотношение возвратов, чем кредиты взятые на такие цели как проведение свадьбы или операции с недвижимостью.

С чем это может быть связано? -Возможно высокий процент займов в целях Автомобиль и Образование связаны с будущим заработком т.е. люди приобретают авто, для работы в Яндекс.Такси например, или берут кредит на обучение, чтобы освоить новую профессию - оба этих займа ведут к улучшению благосостояния заёмщика, что увеличивает соотношение возврата заёмных средств в срок.

В то время как кредит на свадьбу - ничего кроме расходов в себе не несёт и как правило, после свадьбы финансовая нагрузка на заёмщика возрастает (условно, гипотеза, не факт но например - до свадьбы клиент жил у родителей, и многие позиции по финансам закрывали его родители, после свадьбы - эти обязанности уже в новой ячейке общества легли на заёмщика, неспособность точно предвидеть на сколько вырастет нагрузка после образования семьи, приводит к неправильной оценке собственной платежеспособности и соответствующему падению уровня возвратности в данной категории заёмщиков).

С чем связан невысокий уровень возвратности в операциях с жильём сложно сказать по данной выборке, возможно для этого нужен анализ иного характера.


 

# Общий вывод
<a id="id_4"></a>

Полученные данные были избавлены от пропущенных и дублирующихся значений, категоризированы и подготовлены для исследования.
В ходе работы подробно были описаны методы применённые в исследовании, логика проделанных шагов. 

В конце каждого этапа исследования был сделан вывод.

На все поставленные вопросы были даны ответы. 

Также были сделаны рекомендации касательно того, как избежать появления пустых значений в будущем. 


# Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  открыт файл;
- [x]  файл изучен;
- [x]  определены пропущенные значения;
- [x]  заполнены пропущенные значения;
- [x]  есть пояснение, какие пропущенные значения обнаружены;
- [x]  описаны возможные причины появления пропусков в данных;
- [x]  объяснено, по какому принципу заполнены пропуски;
- [x]  заменен вещественный тип данных на целочисленный;
- [x]  есть пояснение, какой метод используется для изменения типа данных и почему;
- [x]  удалены дубликаты;
- [x]  есть пояснение, какой метод используется для поиска и удаления дубликатов;
- [x]  описаны возможные причины появления дубликатов в данных;
- [x]  выделены леммы в значениях столбца с целями получения кредита;
- [x]  описан процесс лемматизации;
- [x]  данные категоризированы;
- [x]  есть объяснение принципа категоризации данных;
- [x]  есть ответ на вопрос: "Есть ли зависимость между наличием детей и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Есть ли зависимость между семейным положением и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Есть ли зависимость между уровнем дохода и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Как разные цели кредита влияют на его возврат в срок?";
- [x]  в каждом этапе есть выводы;
- [x]  есть общий вывод.